content/40.units_and_quantities.md

```diff
@@ -1,269 +1,682 @@
 
-Let's use some common units as examples: gram (`g`), erg (`erg`), and solar mass per cubic megaparsec (`Msun / Mpc`$^3$).
-`g` is an atomic, CGS base unit, `erg` is an atomic unit in CGS, but is not a base unit, and `Msun/Mpc`$^3$ is a combination of atomic units, which are not in CGS, and one of them even has an SI prefix.
-The dimensions of `g` are `mass` and the cgs factor is 1.
-The dimensions of `erg` are `mass * length$^2`$ * `time`$^{-2}$ and the cgs factor is 1.
-The dimensions of `Msun/Mpc`$^3$ are `mass / length`$^3$ and the cgs factor is about 6.8e-41.
+```python
+>>> import unyt as u 
+>>> import numpy as np
+>>> l = np.array([10, 100, 100])*u.Mpc
+>>> print(l.in_cgs())
+[3.08567758e+25 3.08567758e+26 3.08567758e+26] cm
+>>> print(l.in_mks())
+[3.08567758e+23 3.08567758e+24 3.08567758e+24] m
+>>> print(l.in_base("imperial"))
+[1.01236141e+24 1.01236141e+25 1.01236141e+25] ft
+>>> print(l.in_base("geometrized"))
+```
 
-We use the `UnitRegistry` class to define all valid atomic units.
-All unit registries contain a unit symbol lookup table (dict) containing the valid units' dimensionality and cgs conversion factor.
-Here is what it would look like with the above units:
+To convert an array in-place instead of making a copy, use `convert_to_units`:
+
+```python
+>>> momentum_y = sp["gas","mass"]*sp["gas","velocity_y"]
+>>> print(momentum_y)
+[1.45031068e+47 1.40467746e+47 1.42059875e+47 ... 1.01206589e+47
+ 9.94539437e+46 1.06969711e+47] cm*g/s
+>>> momentum_y.convert_to_units("Msun*kpc/Myr")
+>>> print(momentum_y)
+[745945.76032301 722474.98923595 730663.86660289 ... 520541.05697294
+ 511526.58818102 550182.82006007] Msun*kpc/Myr
+```
+
+Naturally, attempting to convert an array to units with a different dimension raises an error (in
+most cases, see below for exceptions):
+
+```python
+>>> print(momentum_y.to("degC"))
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+  File "/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/array.py", line 947, in to
+    return self.in_units(units, equivalence=equivalence, **kwargs)
+           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  File "/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/array.py", line 874, in in_units
+    (conversion_factor, offset) = self.units.get_conversion_factor(
+                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  File "/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/unit_object.py", line 694, in get_conversion_factor
+    return _get_conversion_factor(self, other_units, dtype)
+           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  File "/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/unit_object.py", line 939, in _get_conversion_factor
+    raise UnitConversionError(
+unyt.exceptions.UnitConversionError: Cannot convert between 'Msun*kpc/Myr' (dim '(length)*(mass)/(time)') and '°C' (dim '(temperature)').
+```
```

Comment:
I'm not sure what's happening but the view I'm getting in my browser doesn't seem to understand that this is where the code-block ends, which makes syntax highlighting completely useless in what follows. Maybe those are not the exact correct symbols ?

Reply:
Can you say which browser this happens in? I don't see the same issue.

Reply:
Safari, but it's gone now, so maybe it was just a caching error !